### ISSUE TITLE
Question about submitting form data

### ISSUE THREAD

**Issue by DMGPage** [crwlrsoft/crawler#91]
Hello! 
Nice tool you made, but I have some question, that I want to clarify. In first question here you mentioned something about performing login. But in documentation I can't find any example, how to do that. How to fill input fields and submit specific button. Can you give some examples, how to do that? Or that's still work in progress?

**Comment by otsch** [comment]
Hey @DMGPage 👋

sorry, but even though you can perform any HTTP Request (methods GET, POST, PUT, PATCH, DELETE, with any request headers and/or body you want), I won't ever ship any functionality explicitly dedicated to performing a login.

Crawlers built with this library are meant to behave politely and respect website owners interests (see [politeness features](https://www.crwlr.software/packages/crawler/the-crawler/politeness)). Usually there's a reason that certain data sits behind a login and site-owners wouldn’t like to see bots extracting such data. As I am website owner myself, like most people who are crawling and scraping websites, I don't want to support such practices. There are some rare valid use-cases to perform a login with a crawler, mostly if it's about your own website. But as those are rare and most use-cases are rather unethical, the library won't ever ship with functionality explicitly dedicated to this.

I would also not recommend trying to build such crawlers, because in most cases there are also measures trying to prevent bots from logging in, like CSRF tokens, captchas and measures to identify bots doing it anyways, like honeypot traps. If you're getting caught, you will most likely get blocked by your IP address.

If site owners in some way allow people to collect data that sits behind a login, they'll probably provide an official API to do so, so they'll have control over what data they share with whom. If they don't, fetching such data will probably also be against their terms of use.

In case you're having a legit use-case to log in, you can contact me on [twitter](https://twitter.com/crwlrsoft) or [via the contact form on crwlr.software](https://www.crwlr.software/contact) and maybe I can help you.

**Comment by DMGPage** [comment]
Thank you for answer. Actually I'm not even trying to make something for login. In my case I want to perform search, because many website show result only after performing some search. For example some websites offer some information about company, but don't offer list of all companies. So that if I know, company registration code, I can search and find some result. But for that I need to fill search form and press specific button. In my case it's very common need. Also it's very rare case, that simple search form will be protected by captcha.

**Comment by otsch** [comment]
OK. This is indeed a very common use case. There's two different ways to achieve this:
* You can have a look at the Requests, that are sent when you submit the (search) form, in your browser's developer tools (network tab) and just send the same HTTP Requests in your crawler to get the search results (page).
* Or you can use a headless browser for crawling and actually fill inputs and click buttons.

Personally I usually prefer the first approach and there shouldn't be a problem doing this with the library.
The second one currently isn't really supported. The library can use a headless browser to load pages and return the HTML source code after executing javascript code. But there aren't any steps to perform actions like click, fill inputs, and alike. I plan to add an extension package for this in the future, but this will still take a while.

If you want to give the first approach a try, but need some help, you can send me the actual page and what data you want to extract from it, and I can try to help you.

A third way that most people don't really know about, but is often a very easy solution: 
As you told, the site doesn't have a simple listing of all the companies publicly on it's website, but they still want all company detail pages to be found and indexed by search engines. To achieve this, they provide one or multiple sitemap(s) and link them in their robots.txt file. If that's the case you can find all the company detail pages in their sitemap. Very likely those pages have a unique URL path, like /companies/<some-slug-or-id>, so you can then just crawl all URLs, starting with /companies/... from the sitemap (see https://www.crwlr.software/packages/crawler/v1.0/included-steps/sitemap and https://www.crwlr.software/packages/crawler/v1.0/steps-and-data-flow/output-filters#url-filters).

**Comment by DMGPage** [comment]
> A third way that most people don't really know about, but is often a very easy solution:
As you told, the site doesn't have a simple listing of all the companies publicly on it's website, but they still want all company detail pages to be found and indexed by search engines. To achieve this, they provide one or multiple sitemap(s) and link them in their robots.txt file. If that's the case you can find all the company detail pages in their sitemap.

I most of the time work with automation of government data gathering. And government institutions don't care much about search engines. They just make publicly available data and that's it. Sometimes they even publish data in csv or xlsx, but as government institutions, they often forget to publish something in those files. So we sill have to manually check their website.

But about solution, I found, that if I make my custom step, I can access Symfony Crawler. And they have functionality with form data manipulation. Now only left to put all this together, but it looks promising.

**Comment by otsch** [comment]
Of course, a lot of sites also don't provide a sitemap. Just wanted to mention it, as that's often a solution people don't really think of.

OK, it could probably be difficult to perform a POST request in the step following your custom step with the form data from the previous step 🤔 could probably be useful for the library to make this possible. Maybe something like this:
```php
$crawler
    ->addStep(new MyCustomStepToGenerateFormData()) // This step produces output like ['url' => '...', 'formData' => '...']
    ->addStep(
        Http::post()
            ->useAsUrl('url')
            ->useAsBody('formData')
    );
```

But if you don't have to get something from the search form page, to somehow build/create the data for the POST request, you can just do it like this:
```php
$crawler
    ->addStep(
        Http::post(body: 'search_query=foobar') // or however the data needs to be sent.
    );
```

**Comment by otsch** [comment]
Hey @DMGPage, I hope you were able to solve your problem. I just wanted to inform you, that my suggestion from above - to dynamically get headers and body from previous step's output - is now included in v1.1 of the crawler package. See https://www.crwlr.software/packages/crawler/v1.1/included-steps/http#getting-params-from-previous-step
